Fix #251: check VTablePtr

Indeximal opened the pull request:
My fix for #251 was to do as the comment said, and make sure all VTablePtr must point to an allocated vtable in the language invariant check. Previously this was only done for the metadata, not the pointers themselves.
Additionally, this PR adds trait information for vtablePtr pointer types.

RalfJung commented on tooling/minitest/src/tests/trait_object.rs:
```diff
@@ -318,6 +318,29 @@ fn ub_wrong_vtable_ty() {
     assert_ub::<BasicMem>(p, "Value::Ptr: invalid vtable in metadata");
 }
 
+#[test]
```
Please add a doc comment:
```
/// Ensure we correctly report UB when calling vtable_method_lookup on a pointer
/// that does not point to a vtable.
```

RalfJung commented on spec/lang/representation.md:
```diff
@@ -503,6 +503,9 @@ impl<M: Memory> Machine<M> {
             // However, we do not care about the data stored wherever this pointer points to.
+        } else if let PtrType::VTablePtr = ptr_ty {
+            // Vtable pointers must always point to an allocated vtable
+            self.vtable_from_ptr(ptr.thin_pointer)?;
```
Oh I see, this isn't even about the metadata, it is about the "standalone vtbale pointers" which don't exist in Rust.

RalfJung commented on spec/lang/representation.md:
```diff
@@ -503,6 +501,10 @@ impl<M: Memory> Machine<M> {
 
             // However, we do not care about the data stored wherever this pointer points to.
+        } else if let PtrType::VTablePtr(trait_name) = ptr_ty {
+            // Vtable pointers must always point to an allocated vtable for the correct trait
```
```suggestion
            // This is a "stand-alone" vtable pointer, something that does not exist
            // in surface Rust. Ensure that it points to an allocated vtable for the correct trait.
```

RalfJung commented on spec/lang/representation.md:
```diff
@@ -468,9 +468,7 @@ impl<M: Memory> Machine<M> {
-                // Check that the vtable exists and is for the correct trait
-                let vtable = self.vtable_from_ptr(ptr)?;
-                ensure_else_ub(vtable.trait_name == trait_name, "Value::Ptr: invalid vtable in metadata")?;
+                self.check_pointer(ptr.widen(None), PtrType::VTablePtr(trait_name))?;
```
We are pretty inconsistent with "Ptr" vs "Pointer" it seems...

RalfJung commented on spec/lang/step/operators.md:
```diff
@@ -156,12 +156,8 @@ impl<M: Memory> Machine<M> {
-            // FIXME: Store the TraitName in VTablePtr and have this be a program well-formedness requirement.
-            throw_ub!("the referenced vtable does not have an entry for the invoked method");
-        };
+        // Wellformedness ensures the method is declared
```
```suggestion
        // Wellformedness ensures `ptr` points to a vtable of the right trait,
        // and hence the method exists.
```

RalfJung commented on spec/lang/syntax.md:
```diff
@@ -431,7 +431,7 @@ Finally, the general structure of programs and functions:
 pub struct FnName(pub libspecr::Name);
 pub struct GlobalName(pub libspecr::Name);
-pub struct VTableName(pub libspecr::Name);
+pub struct VTableName(pub TraitName, pub libspecr::Name);
```
This is very odd. A name shouldn't usually contain another name. What is happening here?

Indeximal replied:
The reason is that in the builder API the constant both needs the vtable name and trait name, which is redundant.
But yes, it is weird.

RalfJung replied:
The builder API should never affect what happens in `spec/`.

RalfJung commented on spec/lang/well-formed.md:
```diff
@@ -382,13 +384,16 @@ impl ValueExpr {
-                        // We do not statically have enough information to check that the method name exists in the right vtable,
-                        // but this is only a type safety problem, which we don't catch anyways.
+                        let Some(trait_methods) = prog.traits.get(trait_name) else {
+                            throw_ill_formed!("UnOp::VTableMethodLookup: invalid operand: trait of the vtable pointer doesn't exist");
```
Is this even reachable? I would expect `operand` itself to not be well-formed rather than having a nonsense type.

Indeximal replied:
It was reachable because `PtrType::check_wf` and thus `Type::check_wf` didn't take the program as an argument and thus couldn't check this. Fixed now.

RalfJung commented on spec/lang/well-formed.md:
```diff
@@ -748,6 +753,16 @@ impl Relocation {
 
 impl Program {
     fn check_wf<T: Target>(self) -> Result<()> {
+        // Check vtables: All vtables for the same trait must have the same method names defined.
```
```suggestion
        // Check vtables: All vtables for the same trait must have all trait methods defined.
```

Indeximal commented:
I have reverted the traitname in the vtable name and incorporated other feedback. In particular have I extended `Type::check_wf` to also take the program, to check in both the `TraitObject(trait_name) ` and `PtrType::VTablePtr(trait_name)` that the trait name exists.

Indeximal commented:
Resolved conflicts

RalfJung commented on tooling/minitest/src/tests/trait_object.rs:
```diff
@@ -402,6 +386,37 @@ fn ill_lookup_wrong_ty() {
+        let y = f.declare_local_with_ty(Type::Ptr(PtrType::FnPtr));
+        f.storage_live(y);
+        // Ill formed since the vtable doesn't declare the meth2.
+        // (Actually it might still work, since both miniutil and minimize use low entropy method indices).
```
What do you mean by this?

Indeximal replied:
It is only checked that some method with the same name exists. However, there are a bunch of homonyms of the method because miniutil uses monotonically increasing numbers and minimize indices.

RalfJung replied:
But we control the entire testcase and it's not very big so I fail to see how that's relevant...? Also the test does report "illformed" so how does it make sense to say "it might still work" while at the same time asserting that it does not work?

RalfJung reviewed (commented):
Looks great, thanks! Just some minor comments.

RalfJung commented:
Thanks! Please squash this a little (in particular, "Revert including trait name in vtable name" should not exist, the right representation should be used form the start).

Indeximal commented:
I've fully squashed, makes little sense to spend time untangling the revert specifically.
